<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="downstreamGWAS">
<title>downstreamGWAS • downstreamGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="downstreamGWAS">
<meta property="og:description" content="downstreamGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">downstreamGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/downstreamGWAS.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/meta-analysis.html">meta-analysis</a>
    <a class="dropdown-item" href="../articles/multi-file-datasets.html">multi-file datasets</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>downstreamGWAS</h1>
            
      
      
      <div class="d-none name"><code>downstreamGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/downstreamGWAS/">downstreamGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/" class="external-link">tidyGWAS</a></span><span class="op">)</span></span></code></pre></div>
<p>downstreamGWAS has to ambition make standardized pipelines for
downstream analysis that has genome-wide summary statistics as
<strong><em>primary input</em></strong>. Examples of these types of
analysis are: gene-based tests (with magma or GCTA), or heritability
enrichment (with S-LDSC). The required input for these analyses can
typically be put into one of three categories:</p>
<ol style="list-style-type: decimal">
<li><p>summary statistics</p></li>
<li><p>reference data</p></li>
<li><p>software dependencies</p></li>
</ol>
<p>downstreamGWAS is aimed at making this painless and reproducible.
Each part is tackled in the following way:</p>
<div class="section level4">
<h4 id="summary-statistics">(1) Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h4>
<p>downstreamGWAS relies on summary statistics first being cleaned by
<code><a href="http://arvidharder.com/tidyGWAS/reference/tidyGWAS.html" class="external-link">tidyGWAS::tidyGWAS()</a></code>, and assumes that
<code>output_format="hivestyle"</code> (the default argument), and that
no changes has been made to the resulting folder structure.</p>
<p>With standardized column names and data formats, the functionality to
prepare the summary statistics for the downstream analysis can be
continously reused. For example, the <a href="https://cnsgenomics.com/software/gctb/#Overview" class="external-link">GCTB</a> suite of
genetic analysis uses the .ma format, and downstreamGWAS has defined
method for converting tidyGWAS to .ma in
<code><a href="../reference/to_ma.html">downstreamGWAS::to_ma()</a></code>. No need to write new code each
time!</p>
<p>A common (and very time consuming step) is to first munge summary
statistics, and prepare them with correct filters, column names and
data-types. downstreamGWAS uses the standardized tidyGWAS output to
automate this munging step, for example
<code><a href="../reference/to_ma.html">downstreamGWAS::to_ma()</a></code> or
<code><a href="../reference/to_ldsc.html">downstreamGWAS::to_ldsc()</a></code>.</p>
</div>
<div class="section level4">
<h4 id="reference-data">(2) Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h4>
<p>Another common issue is acquiring and using the same reference data
for analysis. downstreamGWAS uses <a href="https://zenodo.org" class="external-link">Zenodo</a> to host and share reference data
where it is possible. This makes it easy to reproduce your results
across computational clusters and research groups. See for example the
reference files for running stratified LDscore regression <a href="https://zenodo.org/records/8367200" class="external-link">here</a>.</p>
</div>
<div class="section level4">
<h4 id="software-dependencies">(3) Software dependencies<a class="anchor" aria-label="anchor" href="#software-dependencies"></a>
</h4>
<p>Software can be difficult to install properly, especially when you
are working on high performance clusters where you might not have enough
permissions. To handle this, downstreamGWAS makes use of the software
container <a href="https://apptainer.org" class="external-link">apptainer</a>. Apptainer can
use <a href="https://www.docker.com" class="external-link">Docker</a> images to produce a
virtual environment, where sudo privileges are not required. For the
downstream analysis implemented in downstreamGWAS, we have created
docker images for each.</p>
<p>This virtualization also serves as another important factor in
reproductibility, making sure analysis is conducted using the same
underlying software.</p>
</div>
<div class="section level3">
<h3 id="setting-up-downstreamgwas">Setting up downstreamGWAS<a class="anchor" aria-label="anchor" href="#setting-up-downstreamgwas"></a>
</h3>
<p>To illustrate the results, we first create an toy example of a
cleaned tidyGWAS sumstat.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/reference/test_file.html" class="external-link">test_file</a></span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://arvidharder.com/tidyGWAS/reference/tidyGWAS.html" class="external-link">tidyGWAS</a></span><span class="op">(</span></span>
<span>  <span class="co"># Here we input the summary statistics as a data.frame already in R memory</span></span>
<span>  tbl <span class="op">=</span> <span class="va">sumstats</span>, </span>
<span>  <span class="co"># provide the filepath to the refence files you downloaded.</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"example_gwas"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.8.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2024-05-16 11:09:38.676828</span></span>
<span><span class="co">#&gt; with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/Rtmp13I4MG/example_gwas</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">2) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A unique ID is formed by concontenating CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 4a) Validating columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 5) Adding RSID based on CHR:POS. Adding dbSNP based QC flags </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3437 snps matched GRCh38, 9998 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 11 rows with no dbSNP entry</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/Rtmp13I4MG/example_gwas/pipeline_info/removed_nodbsnp.parquet</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 486 rows with missing on either build</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/Rtmp13I4MG/example_gwas/pipeline_info/removed_missing_on_either_build.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">6) Repairing missings statistics columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; missing_on_either_build: 486</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; nodbsnp: 11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 497 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 4.7s</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Saving metadata from analysis to <span style="color: #0000BB;">/tmp/Rtmp13I4MG/example_gwas/metadata.yaml</span></span></span></code></pre></div>
<p>Using only the filepath to output_folder, downstreamGWAS can generate
a slurm script to run the specific downstream analysis. Here for
example, a script to run <a href="https://github.com/zhilizheng/SBayesRC" class="external-link">SbayesRC</a> is created,
ready to run.</p>
<div class="section level4">
<h4 id="running-sbayesrc">Running SbayesRC<a class="anchor" aria-label="anchor" href="#running-sbayesrc"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># downstreamGWAS uses the "HOME" variable to store data on where you keep the reference files between sessions. </span></span>
<span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_envvar.html" class="external-link">local_envvar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"HOME"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dsg_folder</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"downstreamGWAS_folder"</span><span class="op">)</span></span>
<span><span class="co"># use setup to tell downstreamGWAS where the reference data and containers are kept</span></span>
<span><span class="co"># only has to be done once</span></span>
<span><span class="fu"><a href="../reference/setup.html">setup</a></span><span class="op">(</span><span class="va">dsg_folder</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Wrote the downstreamGWAS config file to <span style="color: #0000BB;">/tmp/Rtmp13I4MG/.config/downstreamGWAS/config.yml</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Using <span style="color: #0000BB;">/tmp/Rtmp13I4MG/downstreamGWAS_folder</span> to store downstreamGWAS data</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Reference data required for downstreamGWAS should be stored in <span style="color: #0000BB;">/tmp/Rtmp13I4MG/downstreamGWAS_folder/reference</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Containers required for downstreamGWAS should be stored in <span style="color: #0000BB;">/tmp/Rtmp13I4MG/downstreamGWAS_folder/containers</span></span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_sbayesrc.html">run_sbayesrc</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"example_gwas"</span><span class="op">)</span>,  write_script <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "#!/bin/bash"                                                                                                                                    </span></span>
<span><span class="co">#&gt;  [2] "#SBATCH --mem=8gb"                                                                                                                              </span></span>
<span><span class="co">#&gt;  [3] "#SBATCH --time=24:00:00"                                                                                                                        </span></span>
<span><span class="co">#&gt;  [4] "#SBATCH --output=/tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc/slurm-%j.out"                                                                   </span></span>
<span><span class="co">#&gt;  [5] ""                                                                                                                                               </span></span>
<span><span class="co">#&gt;  [6] "\n"                                                                                                                                             </span></span>
<span><span class="co">#&gt;  [7] "R -e \"downstreamGWAS::to_ma('/tmp/Rtmp13I4MG/example_gwas')\""                                                                                 </span></span>
<span><span class="co">#&gt;  [8] "\n"                                                                                                                                             </span></span>
<span><span class="co">#&gt;  [9] ""                                                                                                                                               </span></span>
<span><span class="co">#&gt; [10] "workdir='/tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc:/mnt'"                                                                                  </span></span>
<span><span class="co">#&gt; [11] "reference_dir='/tmp/Rtmp13I4MG/downstreamGWAS_folder/reference:/src'"                                                                           </span></span>
<span><span class="co">#&gt; [12] "container='/tmp/Rtmp13I4MG/downstreamGWAS_folder/containers/sbayesrc_latest.sif'"                                                               </span></span>
<span><span class="co">#&gt; [13] "code=\"SBayesRC::tidy('/mnt/sumstats.ma',LDdir='/src/gctb/ukbEUR_Imputed',output='/mnt/sumstats.ma')\""                                         </span></span>
<span><span class="co">#&gt; [14] "apptainer exec --cleanenv --bind $workdir,$reference_dir $container R -e \"$code\""                                                             </span></span>
<span><span class="co">#&gt; [15] "\n"                                                                                                                                             </span></span>
<span><span class="co">#&gt; [16] "code=\"SBayesRC::impute('/mnt/sumstats.ma',LDdir='/src/gctb/ukbEUR_Imputed',output='/mnt/sumstats.ma')\""                                       </span></span>
<span><span class="co">#&gt; [17] "apptainer exec --cleanenv --env 'OMP_NUM_THREADS=4' --bind $workdir,$reference_dir $container R -e \"$code\""                                   </span></span>
<span><span class="co">#&gt; [18] "\n"                                                                                                                                             </span></span>
<span><span class="co">#&gt; [19] "code=\"SBayesRC::sbayesrc('/mnt/sumstats.ma',LDdir='/src/gctb/ukbEUR_Imputed',outPrefix='/mnt/sbrc', annot='/src/gctb/annot_baseline2.2.txt')\""</span></span>
<span><span class="co">#&gt; [20] "apptainer exec --cleanenv --env 'OMP_NUM_THREADS=8' --bind $workdir,$reference_dir $container R -e \"$code\""                                   </span></span>
<span><span class="co">#&gt; [21] "\n"                                                                                                                                             </span></span>
<span><span class="co">#&gt; [22] "rm /tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc/*.rds"                                                                                        </span></span>
<span><span class="co">#&gt; [23] "rm /tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc/sumstats.ma*"                                                                                 </span></span>
<span><span class="co">#&gt; [24] "rm /tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc/sbrc_tune_inter.txt*"                                                                         </span></span>
<span><span class="co">#&gt; [25] "gzip /tmp/Rtmp13I4MG/example_gwas/analysis/sbayesrc/sbrc.txt"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-type-analysis-using-s-ldsc">cell-type analysis using S-LDSC<a class="anchor" aria-label="anchor" href="#cell-type-analysis-using-s-ldsc"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_sldsc_cts.html">run_sldsc_cts</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"example_gwas"</span><span class="op">)</span>,cts_file <span class="op">=</span> <span class="st">"toy.cts"</span>, write_script <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "#!/bin/bash"                                                                                                                                                                                                                                                                                                                           </span></span>
<span><span class="co">#&gt;  [2] "#SBATCH --mem=8gb"                                                                                                                                                                                                                                                                                                                     </span></span>
<span><span class="co">#&gt;  [3] "#SBATCH --time=24:00:00"                                                                                                                                                                                                                                                                                                               </span></span>
<span><span class="co">#&gt;  [4] "#SBATCH --output=/tmp/Rtmp13I4MG/example_gwas/analysis/ldsc/sldsc/slurm-%j.out"                                                                                                                                                                                                                                                        </span></span>
<span><span class="co">#&gt;  [5] ""                                                                                                                                                                                                                                                                                                                                      </span></span>
<span><span class="co">#&gt;  [6] "workdir='/tmp/Rtmp13I4MG/example_gwas/analysis/ldsc:/mnt'"                                                                                                                                                                                                                                                                             </span></span>
<span><span class="co">#&gt;  [7] "reference_dir='/tmp/Rtmp13I4MG/downstreamGWAS_folder/reference:/src'"                                                                                                                                                                                                                                                                  </span></span>
<span><span class="co">#&gt;  [8] "container='/tmp/Rtmp13I4MG/downstreamGWAS_folder/containers/ldsc_latest.sif'"                                                                                                                                                                                                                                                          </span></span>
<span><span class="co">#&gt;  [9] "code='python /tools/ldsc/ldsc.py --h2-cts /mnt/ldsc.sumstats.gz --ref-ld-chr /src/sldsc_ref/1000G_EUR_Phase3_baseline/baseline. --ref-ld-chr-cts /src/toy.cts --w-ld-chr /src/sldsc_ref/1000G_Phase3_weights_hm3_no_MHC/weights.hm3_noMHC. --overlap-annot --frqfile-chr /src/sldsc_ref/1000G_Phase3_frq/1000G.EUR.QC. --out /mnt/toy'"</span></span>
<span><span class="co">#&gt; [10] "apptainer exec --cleanenv --bind $workdir,$reference_dir $container $code"</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="setting-up-downstreamgwas-1">Setting up downstreamGWAS<a class="anchor" aria-label="anchor" href="#setting-up-downstreamgwas-1"></a>
</h2>
<p>WORK IN PROGRESS</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
